contrib/aws/aws-sdk-go/aws/aws.go

```diff
@@ -147,3 +163,224 @@ func awsAgent(req *request.Request) string {
 func awsRegion(req *request.Request) string {
 	return req.ClientInfo.SigningRegion
 }
+
+var serviceTags = map[string]func(params interface{}) (map[string]string, error){
+	"sqs":         sqsTags,
+	"s3":          s3Tags,
+	"sns":         snsTags,
+	"dynamodb":    dynamoDBTags,
+	"kinesis":     kinesisTags,
+	"EventBridge": eventBridgeTags,
+	"states":      sfnTags,
+}
+
+func extraTagsForService(req *request.Request) map[string]string {
+	service := awsService(req)
+	fn, ok := serviceTags[service]
+	if !ok {
+		return nil
+	}
+	r, err := fn(req.Params)
+	if err != nil {
+		log.Debug("failed to extract tags for AWS service %q: %v", service, err)
+	}
+	return r
+}
+
+func sqsTags(params interface{}) (map[string]string, error) {
+	var queueURL string
+	switch input := params.(type) {
+	case *sqs.SendMessageInput:
+		queueURL = *input.QueueUrl
+	case *sqs.DeleteMessageInput:
+		queueURL = *input.QueueUrl
+	case *sqs.DeleteMessageBatchInput:
+		queueURL = *input.QueueUrl
+	case *sqs.ReceiveMessageInput:
+		queueURL = *input.QueueUrl
+	case *sqs.SendMessageBatchInput:
+		queueURL = *input.QueueUrl
+	default:
+		return nil, nil
+	}
```

Comment:
Sorry for the delay here. I wanted to spend some more time investigating if there's a way to get around this kind of code. And fortunately, @jonbodner came up with a great idea!!

```
func getField[T any](i interface{}, fieldName string) (T, bool) {
	v := reflect.ValueOf(i)
	for v.Kind() == reflect.Ptr {
		v = v.Elem()
	}
	if v.Kind() != reflect.Struct {
		var zero T
		return zero, false
	}
	out := v.FieldByName(fieldName)
	if !out.IsValid() {
		var zero T
		return zero, false
	}
	outs, ok := out.Interface().(T)
	return outs, ok
}
```

This function then get's much simpler:
```
func queueName(requestInput middleware.InitializeInput) string {
	queueURL, ok := getField[*string](requestInput.Parameters, "QueueUrl")
	if !ok {
		log.Debug("No QueueUrl field in %v", requestInput.Parameters)
		return ""
	}
	parts := strings.Split(*queueURL, "/")
	return parts[len(parts)-1]
}

func bucketName(requestInput middleware.InitializeInput) string {
	ret, ok := getField[*string](requestInput.Parameters, "Bucket")
	if !ok {
		log.Debug("No Bucket field in %v", requestInput.Parameters)
		return ""
	}
	return *ret
}
```

I tested this locally, and it passes!

It's going to be a rework of a lot of the code like this, and may have a performance impact, so we should definitely benchmark it. But I'd really like if we can investigate doing this in a follow-up PR to this one, if we can't do it now. It will make the code a lot less verbose, and more maintainable in the future. We can discuss this offline or in the follow-up PR though.

Reply:
Let's explore this in a future PR, and make sure to benchmark it. No need to do anything in this one.

Reply:
thanks for the suggestion, it looks good! I'll explore this in a follow up PR :)